src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/IRequestExecutionQueue.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+#nullable enable
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public interface IRequestExecutionQueue<RequestContextType>
+{
+    Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(bool mutatesSolutionState, bool requiresLSPSolution, IRequestHandler<TRequestType, TResponseType, RequestContextType> handler, TRequestType? request, string methodName, ILspServices lspServices, CancellationToken cancellationToken);
+
+    Task ExecuteAsync<TRequestType>(bool mutatesSolutionState, bool requiresLSPSolution, INotificationHandler<TRequestType, RequestContextType> handler, TRequestType? request, string methodName, ILspServices lspServices, CancellationToken cancellationToken);
+
+    Task ExecuteAsync(bool mutatesSolutionState, bool requiresLSPSolution, INotificationHandler<RequestContextType> handler, string methodName, ILspServices lspServices, CancellationToken cancellationToken);
+
+    void Start(ILspServices lspServices);
```

Comment:
Why have a separate start / shutdown method vs. starting on construction and shutting down on dispose?

Reply:
For `Start` we need the opportunity to register events onto `RequestServerShutdown` [before starting the queue](https://github.com/dotnet/roslyn/pull/62924/files#diff-68827171a9800c8a40f4ccd71a96a93fb5206420213e3532846c7ab0fa0f453fR151).

The Shutdown part is a good point though, will switch to `IDisposable`.

Reply:
Hmmm, interesting. To Cyrus' point we may want to use iasync disposables and then also make sure that all these bits can be async. The Fire & forget nature of the requestservershutdown bit might be concerning